src/operator/z1.rs

```diff
@@ -110,47 +110,211 @@ where
     }
 }
 
-#[test]
-fn sum_circuit() {
-    let mut z1 = Z1::new(0);
-
-    let expected_result = vec![0, 1, 2, 0, 4, 5];
-
-    // Test `UnaryOperator` API.
-    let mut res = Vec::new();
-    z1.clock_start(0);
-    res.push(z1.eval(&1));
-    res.push(z1.eval(&2));
-    res.push(z1.eval(&3));
-    z1.clock_end(0);
-
-    z1.clock_start(0);
-    res.push(z1.eval_owned(4));
-    res.push(z1.eval_owned(5));
-    res.push(z1.eval_owned(6));
-    z1.clock_end(0);
-
-    assert_eq!(res, expected_result);
-
-    // Test `StrictUnaryOperator` API.
-    let mut res = Vec::new();
-    z1.clock_start(0);
-    res.push(z1.get_output());
-    z1.eval_strict(&1);
-    res.push(z1.get_output());
-    z1.eval_strict(&2);
-    res.push(z1.get_output());
-    z1.eval_strict(&3);
-    z1.clock_end(0);
-
-    z1.clock_start(0);
-    res.push(z1.get_output());
-    z1.eval_strict_owned(4);
-    res.push(z1.get_output());
-    z1.eval_strict_owned(5);
-    res.push(z1.get_output());
-    z1.eval_strict_owned(6);
-    z1.clock_end(0);
-
-    assert_eq!(res, expected_result);
+/// z^-1 operator over streams of streams.
+///
+/// The operator stores a complete nested stream consumed at the last iteration
+/// of the parent clock and outputs it at the next parent clock cycle
+/// It outputs a stream of zeros value in the first parent clock tick.
+///
+/// It is a strict operator.
+///
+/// # Examples
+///
+/// Input (one row per parent timestamps):
+///
+/// ```text
+/// 1 2 3 4
+/// 1 1 1 1 1
+/// 2 2 2 0 0
+/// ```
+///
+/// Output:
+///
+/// ```text
+/// 0 0 0 0
+/// 1 2 3 4 0
+/// 1 1 1 1 1
+/// ```
+pub struct Z1Nested<T> {
+    zero: T,
+    timestamp: usize,
+    val: Vec<T>,
+}
+
+impl<T> Z1Nested<T>
+where
+    T: Clone,
+{
+    pub fn new(zero: T) -> Self {
+        Self {
+            zero,
+            timestamp: 0,
+            val: vec![],
+        }
+    }
+
+    fn reset(&mut self) {
+        self.timestamp = 0;
+        self.val = vec![];
```

Comment:
Keep this buffer allocated
```suggestion
        self.val.clear();
```

Reply:
This is reset for the outer timestamp, basically the whole parent circuit starting from scratch, so there's no reason to expect that we'll need similar amount of inner iterations.  In any case, savings are negligible across many parent clock ticks.

Reply:
I mean, there's no reason not to keep the memory allocated, it's only a benefit really